src/main/java/com/google/cloud/pontem/FormatGenericRecordAsSpannerMutationFn.java

```diff
@@ -0,0 +1,290 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.pontem;
+
+import com.google.cloud.ByteArray;
+import com.google.cloud.Date;
+import com.google.cloud.Timestamp;
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.Type;
+import com.google.cloud.spanner.Type.Code;
+import com.google.common.collect.ImmutableMap;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.util.Utf8;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+
+/**
+ * A {@code SimpleFunction} that converts a {@code GenericRecord} to a {@code Mutation}.
+ *
+ * <p>See https://cloud.google.com/spanner/docs/data-types
+ */
+public class FormatGenericRecordAsSpannerMutationFn
+    extends SimpleFunction<GenericRecord, Mutation> {
+
+  private final Counter tableRowsReadCounter;
+  private final String tableName;
+  private final ImmutableMap<String, Type> mapOfColumnNamesToSpannerTypes;
+
+  FormatGenericRecordAsSpannerMutationFn(
+      String tableName, ImmutableMap<String, Type> mapOfColumnNamesToSpannerTypes) {
+    this.tableRowsReadCounter =
+        Metrics.counter(FormatGenericRecordAsSpannerMutationFn.class, "recordsRead_" + tableName);
+    this.tableName = tableName;
+    this.mapOfColumnNamesToSpannerTypes = mapOfColumnNamesToSpannerTypes;
+  }
+
+  @Override
+  public Mutation apply(GenericRecord inputRow) {
+    Mutation.WriteBuilder mutationBuilder = Mutation.newInsertOrUpdateBuilder(this.tableName);
+
+    Schema tableSchema = inputRow.getSchema();
+    List<Schema.Field> tableFields = tableSchema.getFields();
+    for (Schema.Field tableField : tableFields) {
+
+      String fieldName = tableField.name();
+      Type spannerFieldType = this.mapOfColumnNamesToSpannerTypes.get(fieldName);
+      Schema columnSchema = tableField.schema();
+      // Since the Avro column schema could be a UNION of types (e.g., [STRING, NULL]),
+      // we need to unpack that UNION to simply be a noramlly recognized Avro type (e.g., STRING).
+      if (columnSchema.getType() == Schema.Type.UNION) {
+        columnSchema = AvroUtil.getSingleAvroTypeFromNullableUnion(columnSchema);
+      }
+      Schema.Type avroFieldType = columnSchema.getType();
+
+      if (spannerFieldType.getCode() == Code.STRING && avroFieldType == Schema.Type.STRING) {
```

Comment:
A future nice to have specially if we'll keep adding data types would be to create a chain of responsibility to handle these kinds of type dependent parsers